// @cyclonedx/cyclonedx-npm@6.0.0/dist/npmRunner.js
"use strict";
/*!
This file is part of CycloneDX generator for NPM projects.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

SPDX-License-Identifier: Apache-2.0
Copyright (c) OWASP Foundation. All Rights Reserved.
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.NpmRunner = void 0;
const node_child_process_1 = require("node:child_process");
const node_fs_1 = require("node:fs");
const node_os_1 = require("node:os");
const node_path_1 = require("node:path");
class NpmRunner {
    static #jsMatcher = /\.[cm]?js$/;
    static #npxMatcher = /(^|\\|\/)npx-cli\.js$/;
    run;
    constructor(process_, console_) {
        this.run = NpmRunner.#makeNpmRunner(process_, console_);
    }
    #version;
    getVersion(options = {}) {
        if (this.#version === undefined) {
            this.#version = this.run(['--version'], {
                ...options,
                stdio: ['ignore', 'pipe', 'ignore'],
                encoding: 'utf-8'
            }).trim();
        }
        return this.#version;
    }
    static #getExecPathEnv(process_, console_) {
        console_.debug('DEBUG | looking up env NPM...');
        let npmPath = process_.env.npm_execpath ?? '';
        if (npmPath === '') {
            console_.debug('DEBUG | env NPM empty');
            return undefined;
        }
        if (NpmRunner.#npxMatcher.test(npmPath)) {
            console_.debug('DEBUG | command: npx-cli.js usage detected, checking for npm-cli.js ...');
            npmPath = (0, node_path_1.resolve)(npmPath.replace(NpmRunner.#npxMatcher, '$1npm-cli.js'));
        }
        if (!(0, node_fs_1.existsSync)(npmPath)) {
            throw new Error(`Environment variable "npm_execpath" is set, but the path does not exist: ${JSON.stringify(npmPath)}`);
        }
        console_.debug('DEBUG | env NPM found: %s', npmPath);
        return npmPath;
    }
    static #getExecPathSys(process_, console_) {
        console_.debug('DEBUG | looking up system NPM...');
        let npmPath = '';
        const tmpDir = (0, node_fs_1.mkdtempSync)((0, node_path_1.join)((0, node_os_1.tmpdir)(), 'cyclonedx-npm_execpath-'));
        const packageManifest = (0, node_path_1.join)(tmpDir, 'package.json');
        try {
            const packageManifestFH = (0, node_fs_1.openSync)(packageManifest, 'w');
            try {
                (0, node_fs_1.writeFileSync)(packageManifestFH, JSON.stringify({
                    'private': true,
                    'name': '@cyclonedx/cyclonedx-npm_execpath',
                    'scripts': {
                        'npm_execpath': 'node -p process.env.npm_execpath'
                    }
                }));
            }
            finally {
                (0, node_fs_1.closeSync)(packageManifestFH);
            }
            npmPath = (0, node_child_process_1.execSync)('npm run --silent npm_execpath', {
                cwd: tmpDir,
                env: process_.env,
                windowsHide: true,
                stdio: ['ignore', 'pipe', 'ignore'],
                encoding: 'utf-8'
            }).trim();
        }
        catch (err) {
            throw new Error('Failed looking up system NPM', { cause: err });
        }
        finally {
            (0, node_fs_1.rmSync)(packageManifest, { force: true });
            (0, node_fs_1.rmSync)(tmpDir, { recursive: true, force: true });
        }
        if (npmPath === '' || !(0, node_fs_1.existsSync)(npmPath)) {
            throw new Error(`Missing system NPM ${JSON.stringify(npmPath)}`);
        }
        console_.debug('DEBUG | system NPM found: %s', npmPath);
        return npmPath;
    }
    static #makeNpmRunner(process_, console_) {
        const npmJsPath = NpmRunner.#getExecPathEnv(process_, console_)
            ?? NpmRunner.#getExecPathSys(process_, console_);
        if (!NpmRunner.#jsMatcher.test(npmJsPath)) {
            throw new Error(`Unexpected npmJsPath ${JSON.stringify(npmJsPath)}`);
        }
        const nodeExecPath = process_.execPath;
        console_.debug('DEBUG | makeNpmRunner caused execFileSync "%s" with "-- %s"', nodeExecPath, npmJsPath);
        return ((args, options) => (0, node_child_process_1.execFileSync)(nodeExecPath, ['--', npmJsPath, ...args], { ...(options ?? {}), shell: false, windowsHide: true }));
    }
}
exports.NpmRunner = NpmRunner;
